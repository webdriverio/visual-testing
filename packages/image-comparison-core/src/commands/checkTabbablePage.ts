import drawTabbableOnCanvas from '../clientSideScripts/drawTabbableOnCanvas.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import checkFullPageScreen from './checkFullPageScreen.js'
import type { ImageCompareResult } from '../index.js'
import type { InternalCheckTabbablePageMethodOptions } from './check.interfaces.js'

/**
 * Compare an image with all tab executions
 */
export default async function checkTabbablePage(
    {
        browserInstance,
        checkTabbableOptions,
        folders,
        instanceData,
        isNativeContext = false,
        tag,
        testContext,
    }: InternalCheckTabbablePageMethodOptions
): Promise<ImageCompareResult | number> {
    // 1a. Check if the method is supported in native context
    if (isNativeContext) {
        throw new Error('The method checkTabbablePage is not supported in native context for native mobile apps!')
    }

    // 1b. Inject drawing the tabbables
    await browserInstance.execute(drawTabbableOnCanvas, checkTabbableOptions.wic.tabbableOptions)

    // 2. Create the screenshot
    const fullPageCompareData = await checkFullPageScreen({
        browserInstance,
        checkFullPageOptions: checkTabbableOptions,
        instanceData,
        folders,
        isNativeContext,
        testContext,
        tag,
    })

    // 3. Remove the canvas
    await browserInstance.execute(removeElementFromDom, 'wic-tabbable-canvas')

    // 4. Return the data
    return fullPageCompareData
}
