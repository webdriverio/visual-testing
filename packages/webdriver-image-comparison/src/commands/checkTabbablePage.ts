import drawTabbableOnCanvas from '../clientSideScripts/drawTabbableOnCanvas.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import checkFullPageScreen from './checkFullPageScreen.js'
import type { ImageCompareResult } from 'src/index.js'
import type { InternalCheckTabbablePageMethodOptions } from './check.interfaces.js'

/**
 * Compare an image with all tab executions
 */
export default async function checkTabbablePage(
    {
        methods,
        instanceData,
        folders,
        tag,
        checkTabbableOptions,
        isEmulated = false,
        isNativeContext = false,
        testContext,
    }: InternalCheckTabbablePageMethodOptions
): Promise<ImageCompareResult | number> {
    // 1a. Check if the method is supported in native context
    if (isNativeContext) {
        throw new Error('The method checkTabbablePage is not supported in native context for native mobile apps!')
    }

    // 1b. Inject drawing the tabbables
    await methods.executor(drawTabbableOnCanvas, checkTabbableOptions.wic.tabbableOptions)

    // 2. Create the screenshot
    const fullPageCompareData = await checkFullPageScreen({
        methods,
        instanceData,
        folders,
        tag,
        checkFullPageOptions:
        checkTabbableOptions,
        isNativeContext,
        isEmulated,
        testContext,
    })

    // 3. Remove the canvas
    await methods.executor(removeElementFromDom, 'wic-tabbable-canvas')

    // 4. Return the data
    return fullPageCompareData
}
