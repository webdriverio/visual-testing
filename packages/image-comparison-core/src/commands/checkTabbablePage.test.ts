import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import checkTabbablePage from './checkTabbablePage.js'
import type { InternalCheckTabbablePageMethodOptions } from './check.interfaces.js'
import { BASE_CHECK_OPTIONS } from '../mocks/mocks.js'

vi.mock('../clientSideScripts/drawTabbableOnCanvas.js', () => ({
    default: vi.fn()
}))
vi.mock('../clientSideScripts/removeElementFromDom.js', () => ({
    default: vi.fn()
}))
vi.mock('./checkFullPageScreen.js', () => ({
    default: vi.fn().mockResolvedValue({
        fileName: 'test-tabbable.png',
        misMatchPercentage: 0,
        isExactSameImage: true,
        isNewBaseline: false,
        isAboveTolerance: false,
    })
}))

describe('checkTabbablePage', () => {
    let checkFullPageScreenSpy: ReturnType<typeof vi.fn>

    const baseOptions: InternalCheckTabbablePageMethodOptions = {
        checkTabbableOptions: {
            wic: {
                ...BASE_CHECK_OPTIONS.wic,
                tabbableOptions: {
                    circle: {
                        backgroundColor: 'red',
                        borderColor: 'blue',
                        borderWidth: 2,
                        fontColor: 'white',
                        fontFamily: 'Arial',
                        fontSize: 10,
                        size: 10,
                        showNumber: true,
                    },
                    line: {
                        color: 'green',
                        width: 2,
                    },
                }
            },
            method: {
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                enableLegacyScreenshotMethod: false,
                fullPageScrollTimeout: 1500,
                hideAfterFirstScroll: [],
                hideScrollBars: true,
                hideElements: [],
                removeElements: [],
                waitForFontsLoaded: true,
            }
        },
        browserInstance: {
            isAndroid: false,
            execute: vi.fn().mockResolvedValue(undefined)
        } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        isNativeContext: false,
        tag: 'test-tabbable',
        testContext: BASE_CHECK_OPTIONS.testContext
    }

    beforeEach(async () => {
        const checkFullPageScreen = (await import('./checkFullPageScreen.js')).default
        checkFullPageScreenSpy = vi.mocked(checkFullPageScreen)
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should throw error when used in native context', async () => {
        const options = {
            ...baseOptions,
            isNativeContext: true
        }

        await expect(checkTabbablePage(options)).rejects.toThrow(
            'The method checkTabbablePage is not supported in native context for native mobile apps!'
        )
    })

    it('should execute checkTabbablePage with basic options', async () => {
        const result = await checkTabbablePage(baseOptions)

        expect(result).toMatchSnapshot()
        expect(baseOptions.browserInstance.execute).toMatchSnapshot()
        expect(checkFullPageScreenSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle Android device correctly', async () => {
        const options = {
            ...baseOptions,
            browserInstance: {
                isAndroid: true,
                execute: vi.fn().mockResolvedValue(undefined)
            } as any,
            instanceData: {
                ...baseOptions.instanceData,
                deviceName: 'Pixel 4',
                isAndroid: true,
                isIOS: false,
                platformName: 'Android',
                platformVersion: '11.0'
            },
            testContext: {
                ...baseOptions.testContext,
                instanceData: {
                    ...baseOptions.testContext.instanceData,
                    deviceName: 'Pixel 4',
                    platform: { name: 'Android', version: '11.0' },
                    isAndroid: true,
                    isIOS: false
                }
            }
        }

        await checkTabbablePage(options)

        expect(checkFullPageScreenSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle custom tabbable options', async () => {
        const options = {
            ...baseOptions,
            checkTabbableOptions: {
                ...baseOptions.checkTabbableOptions,
                wic: {
                    ...baseOptions.checkTabbableOptions.wic,
                    tabbableOptions: {
                        circle: {
                            backgroundColor: 'yellow',
                            borderColor: 'black',
                            borderWidth: 3,
                            fontColor: 'black',
                            fontFamily: 'Helvetica',
                            fontSize: 12,
                            size: 15,
                            showNumber: false,
                        },
                        line: {
                            color: 'red',
                            width: 3,
                        },
                    }
                }
            }
        }

        await checkTabbablePage(options)

        expect(baseOptions.browserInstance.execute).toMatchSnapshot()
    })

    it('should handle errors during execution', async () => {
        const options = {
            ...baseOptions,
            browserInstance: {
                ...baseOptions.browserInstance,
                execute: vi.fn().mockRejectedValue(new Error('Execution failed')),
                sessionStatus: vi.fn(),
                sessionNew: vi.fn(),
                sessionEnd: vi.fn(),
                sessionSubscribe: vi.fn(),
            } as any
        }

        await expect(checkTabbablePage(options)).rejects.toThrow('Execution failed')
    })

    it('should handle hybrid app options correctly', async () => {
        const options = {
            ...baseOptions,
            checkTabbableOptions: {
                ...baseOptions.checkTabbableOptions,
                wic: {
                    ...baseOptions.checkTabbableOptions.wic,
                    isHybridApp: true
                }
            }
        }

        await checkTabbablePage(options)

        expect(checkFullPageScreenSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle default tabbable options', async () => {
        const options = {
            ...baseOptions,
            checkTabbableOptions: {
                ...baseOptions.checkTabbableOptions,
                wic: {
                    ...baseOptions.checkTabbableOptions.wic,
                    tabbableOptions: {
                        circle: {
                            backgroundColor: 'red',
                            borderColor: 'blue',
                            borderWidth: 2,
                            fontColor: 'white',
                            fontFamily: 'Arial',
                            fontSize: 10,
                            size: 10,
                            showNumber: true,
                        },
                        line: {
                            color: 'green',
                            width: 2,
                        },
                    }
                }
            }
        }

        await checkTabbablePage(options)

        expect(baseOptions.browserInstance.execute).toMatchSnapshot()
    })
})
