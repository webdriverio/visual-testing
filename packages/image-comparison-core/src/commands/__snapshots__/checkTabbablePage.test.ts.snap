// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkTabbablePage > should execute checkTabbablePage with basic options 1`] = `
{
  "fileName": "test-tabbable.png",
  "isAboveTolerance": false,
  "isExactSameImage": true,
  "isNewBaseline": false,
  "misMatchPercentage": 0,
}
`;

exports[`checkTabbablePage > should execute checkTabbablePage with basic options 2`] = `
[MockFunction spy] {
  "calls": [
    [
      [MockFunction spy],
      {
        "circle": {
          "backgroundColor": "red",
          "borderColor": "blue",
          "borderWidth": 2,
          "fontColor": "white",
          "fontFamily": "Arial",
          "fontSize": 10,
          "showNumber": true,
          "size": 10,
        },
        "line": {
          "color": "green",
          "width": 2,
        },
      },
    ],
    [
      [MockFunction spy],
      "wic-tabbable-canvas",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`checkTabbablePage > should execute checkTabbablePage with basic options 3`] = `
[
  {
    "browserInstance": {
      "execute": [MockFunction spy] {
        "calls": [
          [
            [MockFunction spy],
            {
              "circle": {
                "backgroundColor": "red",
                "borderColor": "blue",
                "borderWidth": 2,
                "fontColor": "white",
                "fontFamily": "Arial",
                "fontSize": 10,
                "showNumber": true,
                "size": 10,
              },
              "line": {
                "color": "green",
                "width": 2,
              },
            },
          ],
          [
            [MockFunction spy],
            "wic-tabbable-canvas",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "isAndroid": false,
    },
    "checkFullPageOptions": {
      "method": {
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "fullPageScrollTimeout": 1500,
        "hideAfterFirstScroll": [],
        "hideElements": [],
        "hideScrollBars": true,
        "removeElements": [],
        "waitForFontsLoaded": true,
      },
      "wic": {
        "addIOSBezelCorners": false,
        "addressBarShadowPadding": 6,
        "autoElementScroll": true,
        "autoSaveBaseline": false,
        "clearFolder": false,
        "compareOptions": {
          "blockOutSideBar": false,
          "blockOutStatusBar": false,
          "blockOutToolBar": false,
          "createJsonReportFiles": false,
          "diffPixelBoundingBoxProximity": 5,
          "ignoreAlpha": false,
          "ignoreAntialiasing": false,
          "ignoreColors": false,
          "ignoreLess": false,
          "ignoreNothing": false,
          "rawMisMatchPercentage": false,
          "returnAllCompareData": false,
          "saveAboveTolerance": 0,
          "scaleImagesToSameSize": false,
        },
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
        "fullPageScrollTimeout": 1500,
        "hideScrollBars": true,
        "isHybridApp": false,
        "savePerInstance": true,
        "tabbableOptions": {
          "circle": {
            "backgroundColor": "red",
            "borderColor": "blue",
            "borderWidth": 2,
            "fontColor": "white",
            "fontFamily": "Arial",
            "fontSize": 10,
            "showNumber": true,
            "size": 10,
          },
          "line": {
            "color": "green",
            "width": 2,
          },
        },
        "toolBarShadowPadding": 6,
        "userBasedFullPageScreenshot": false,
        "waitForFontsLoaded": true,
      },
    },
    "folders": {
      "actualFolder": "/test/actual",
      "baselineFolder": "/test/baseline",
      "diffFolder": "/test/diff",
    },
    "instanceData": {
      "appName": "TestApp",
      "browserName": "Chrome",
      "browserVersion": "118.0.0.0",
      "deviceName": "iPhone 14",
      "devicePixelRatio": 2,
      "deviceRectangles": {
        "bottomBar": {
          "height": 0,
          "width": 390,
          "x": 0,
          "y": 800,
        },
        "homeBar": {
          "height": 34,
          "width": 390,
          "x": 0,
          "y": 780,
        },
        "leftSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "rightSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "screenSize": {
          "height": 844,
          "width": 390,
        },
        "statusBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "statusBarAndAddressBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "viewport": {
          "height": 733,
          "width": 390,
          "x": 0,
          "y": 47,
        },
      },
      "initialDevicePixelRatio": 2,
      "isAndroid": false,
      "isIOS": true,
      "isMobile": true,
      "logName": "test-log",
      "name": "test-device",
      "nativeWebScreenshot": false,
      "platformName": "iOS",
      "platformVersion": "17.0",
    },
    "isNativeContext": false,
    "tag": "test-tabbable",
    "testContext": {
      "commandName": "checkScreen",
      "framework": "vitest",
      "instanceData": {
        "app": "TestApp",
        "browser": {
          "name": "Chrome",
          "version": "118.0.0.0",
        },
        "deviceName": "iPhone 14",
        "isAndroid": false,
        "isIOS": true,
        "isMobile": true,
        "platform": {
          "name": "iOS",
          "version": "17.0",
        },
      },
      "parent": "test suite",
      "tag": "test-tag",
      "title": "test title",
    },
  },
]
`;

exports[`checkTabbablePage > should handle Android device correctly 1`] = `
[
  {
    "browserInstance": {
      "execute": [MockFunction spy] {
        "calls": [
          [
            [MockFunction spy],
            {
              "circle": {
                "backgroundColor": "red",
                "borderColor": "blue",
                "borderWidth": 2,
                "fontColor": "white",
                "fontFamily": "Arial",
                "fontSize": 10,
                "showNumber": true,
                "size": 10,
              },
              "line": {
                "color": "green",
                "width": 2,
              },
            },
          ],
          [
            [MockFunction spy],
            "wic-tabbable-canvas",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "isAndroid": true,
    },
    "checkFullPageOptions": {
      "method": {
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "fullPageScrollTimeout": 1500,
        "hideAfterFirstScroll": [],
        "hideElements": [],
        "hideScrollBars": true,
        "removeElements": [],
        "waitForFontsLoaded": true,
      },
      "wic": {
        "addIOSBezelCorners": false,
        "addressBarShadowPadding": 6,
        "autoElementScroll": true,
        "autoSaveBaseline": false,
        "clearFolder": false,
        "compareOptions": {
          "blockOutSideBar": false,
          "blockOutStatusBar": false,
          "blockOutToolBar": false,
          "createJsonReportFiles": false,
          "diffPixelBoundingBoxProximity": 5,
          "ignoreAlpha": false,
          "ignoreAntialiasing": false,
          "ignoreColors": false,
          "ignoreLess": false,
          "ignoreNothing": false,
          "rawMisMatchPercentage": false,
          "returnAllCompareData": false,
          "saveAboveTolerance": 0,
          "scaleImagesToSameSize": false,
        },
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
        "fullPageScrollTimeout": 1500,
        "hideScrollBars": true,
        "isHybridApp": false,
        "savePerInstance": true,
        "tabbableOptions": {
          "circle": {
            "backgroundColor": "red",
            "borderColor": "blue",
            "borderWidth": 2,
            "fontColor": "white",
            "fontFamily": "Arial",
            "fontSize": 10,
            "showNumber": true,
            "size": 10,
          },
          "line": {
            "color": "green",
            "width": 2,
          },
        },
        "toolBarShadowPadding": 6,
        "userBasedFullPageScreenshot": false,
        "waitForFontsLoaded": true,
      },
    },
    "folders": {
      "actualFolder": "/test/actual",
      "baselineFolder": "/test/baseline",
      "diffFolder": "/test/diff",
    },
    "instanceData": {
      "appName": "TestApp",
      "browserName": "Chrome",
      "browserVersion": "118.0.0.0",
      "deviceName": "Pixel 4",
      "devicePixelRatio": 2,
      "deviceRectangles": {
        "bottomBar": {
          "height": 0,
          "width": 390,
          "x": 0,
          "y": 800,
        },
        "homeBar": {
          "height": 34,
          "width": 390,
          "x": 0,
          "y": 780,
        },
        "leftSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "rightSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "screenSize": {
          "height": 844,
          "width": 390,
        },
        "statusBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "statusBarAndAddressBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "viewport": {
          "height": 733,
          "width": 390,
          "x": 0,
          "y": 47,
        },
      },
      "initialDevicePixelRatio": 2,
      "isAndroid": true,
      "isIOS": false,
      "isMobile": true,
      "logName": "test-log",
      "name": "test-device",
      "nativeWebScreenshot": false,
      "platformName": "Android",
      "platformVersion": "11.0",
    },
    "isNativeContext": false,
    "tag": "test-tabbable",
    "testContext": {
      "commandName": "checkScreen",
      "framework": "vitest",
      "instanceData": {
        "app": "TestApp",
        "browser": {
          "name": "Chrome",
          "version": "118.0.0.0",
        },
        "deviceName": "Pixel 4",
        "isAndroid": true,
        "isIOS": false,
        "isMobile": true,
        "platform": {
          "name": "Android",
          "version": "11.0",
        },
      },
      "parent": "test suite",
      "tag": "test-tag",
      "title": "test title",
    },
  },
]
`;

exports[`checkTabbablePage > should handle custom tabbable options 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [MockFunction spy],
      {
        "circle": {
          "backgroundColor": "yellow",
          "borderColor": "black",
          "borderWidth": 3,
          "fontColor": "black",
          "fontFamily": "Helvetica",
          "fontSize": 12,
          "showNumber": false,
          "size": 15,
        },
        "line": {
          "color": "red",
          "width": 3,
        },
      },
    ],
    [
      [MockFunction spy],
      "wic-tabbable-canvas",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`checkTabbablePage > should handle default tabbable options 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [MockFunction spy],
      {
        "circle": {
          "backgroundColor": "red",
          "borderColor": "blue",
          "borderWidth": 2,
          "fontColor": "white",
          "fontFamily": "Arial",
          "fontSize": 10,
          "showNumber": true,
          "size": 10,
        },
        "line": {
          "color": "green",
          "width": 2,
        },
      },
    ],
    [
      [MockFunction spy],
      "wic-tabbable-canvas",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`checkTabbablePage > should handle hybrid app options correctly 1`] = `
[
  {
    "browserInstance": {
      "execute": [MockFunction spy] {
        "calls": [
          [
            [MockFunction spy],
            {
              "circle": {
                "backgroundColor": "red",
                "borderColor": "blue",
                "borderWidth": 2,
                "fontColor": "white",
                "fontFamily": "Arial",
                "fontSize": 10,
                "showNumber": true,
                "size": 10,
              },
              "line": {
                "color": "green",
                "width": 2,
              },
            },
          ],
          [
            [MockFunction spy],
            "wic-tabbable-canvas",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "isAndroid": false,
    },
    "checkFullPageOptions": {
      "method": {
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "fullPageScrollTimeout": 1500,
        "hideAfterFirstScroll": [],
        "hideElements": [],
        "hideScrollBars": true,
        "removeElements": [],
        "waitForFontsLoaded": true,
      },
      "wic": {
        "addIOSBezelCorners": false,
        "addressBarShadowPadding": 6,
        "autoElementScroll": true,
        "autoSaveBaseline": false,
        "clearFolder": false,
        "compareOptions": {
          "blockOutSideBar": false,
          "blockOutStatusBar": false,
          "blockOutToolBar": false,
          "createJsonReportFiles": false,
          "diffPixelBoundingBoxProximity": 5,
          "ignoreAlpha": false,
          "ignoreAntialiasing": false,
          "ignoreColors": false,
          "ignoreLess": false,
          "ignoreNothing": false,
          "rawMisMatchPercentage": false,
          "returnAllCompareData": false,
          "saveAboveTolerance": 0,
          "scaleImagesToSameSize": false,
        },
        "disableBlinkingCursor": false,
        "disableCSSAnimation": false,
        "enableLayoutTesting": false,
        "enableLegacyScreenshotMethod": false,
        "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
        "fullPageScrollTimeout": 1500,
        "hideScrollBars": true,
        "isHybridApp": true,
        "savePerInstance": true,
        "tabbableOptions": {
          "circle": {
            "backgroundColor": "red",
            "borderColor": "blue",
            "borderWidth": 2,
            "fontColor": "white",
            "fontFamily": "Arial",
            "fontSize": 10,
            "showNumber": true,
            "size": 10,
          },
          "line": {
            "color": "green",
            "width": 2,
          },
        },
        "toolBarShadowPadding": 6,
        "userBasedFullPageScreenshot": false,
        "waitForFontsLoaded": true,
      },
    },
    "folders": {
      "actualFolder": "/test/actual",
      "baselineFolder": "/test/baseline",
      "diffFolder": "/test/diff",
    },
    "instanceData": {
      "appName": "TestApp",
      "browserName": "Chrome",
      "browserVersion": "118.0.0.0",
      "deviceName": "iPhone 14",
      "devicePixelRatio": 2,
      "deviceRectangles": {
        "bottomBar": {
          "height": 0,
          "width": 390,
          "x": 0,
          "y": 800,
        },
        "homeBar": {
          "height": 34,
          "width": 390,
          "x": 0,
          "y": 780,
        },
        "leftSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "rightSidePadding": {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "screenSize": {
          "height": 844,
          "width": 390,
        },
        "statusBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "statusBarAndAddressBar": {
          "height": 47,
          "width": 390,
          "x": 0,
          "y": 0,
        },
        "viewport": {
          "height": 733,
          "width": 390,
          "x": 0,
          "y": 47,
        },
      },
      "initialDevicePixelRatio": 2,
      "isAndroid": false,
      "isIOS": true,
      "isMobile": true,
      "logName": "test-log",
      "name": "test-device",
      "nativeWebScreenshot": false,
      "platformName": "iOS",
      "platformVersion": "17.0",
    },
    "isNativeContext": false,
    "tag": "test-tabbable",
    "testContext": {
      "commandName": "checkScreen",
      "framework": "vitest",
      "instanceData": {
        "app": "TestApp",
        "browser": {
          "name": "Chrome",
          "version": "118.0.0.0",
        },
        "deviceName": "iPhone 14",
        "isAndroid": false,
        "isIOS": true,
        "isMobile": true,
        "platform": {
          "name": "iOS",
          "version": "17.0",
        },
      },
      "parent": "test suite",
      "tag": "test-tag",
      "title": "test title",
    },
  },
]
`;
